Patches for OpenBLAS 0.3.12 to support NVIDIA HPC SDK Compilers release 20.11 and later

Good news!  I have been working with our developers to clean up a number of the residual issues involved with compiling OpenBLAS with the NVIDIA HPC SDK Compilers, and as of the 20.11 release (due out shortly) - there are only a small number of changes needed against the OpenBLAS 0.3.12 source tree to allow it to compile it out of the box with our compilers.

Summary of these changes:

1. `-Mnollvm` is no longer needed to compile OpenBLAS with our compilers.  In fact, the old NoLLVM compilers do not even exist anymore.  This flag no longer has any effect and should be removed from the OpenBLAS build with modern releases of our compilers.

2. The `-D__MMX__` flag used to work around a previous issue is also no longer needed.  That issue has been fixed.

3. `f_check` needs a small tweak to correctly detect `nvfortran` as a PGI variant, and thus switch in the appropriate calling conventions for complex numbers.

4. In order to compile kernels that use `_mm256` and `_mm512` intrinsics with our compilers, you have to pass the `-mavx512f` flag.  Otherwise, these intrinsics are not enabled in our compilers by default.  I have modified the SKYLAKEX and COOPERLAKE build rules in `kernel/Makefile` to pass this flag.  I also tested this flag with GCC 10.2.0, and confirmed that it does not adversely affect compiling OpenBLAS with GCC.

Note I built OpenBLAS 0.3.12 with these patches with our 20.11 compilers as follows:

`make CC=nvc FC=nvfortran HOSTCC=gcc DYNAMIC_ARCH=1 USE_OPENMP=1 NUM_THREADS=512 all`

Please give these patches a test with our 20.11 compilers once they are released, and if you agree they work well for you, consider integrating them into the OpenBLAS source tree ahead of the next OpenBLAS release.  I think they will greatly help our customers who also use OpenBLAS.  Cheers!

[openblas-nvhpc.patch.txt](https://github.com/xianyi/OpenBLAS/files/5639534/openblas-nvhpc.patch.txt)

Does the same patch apply to 20.7 or 20.9?

Edit: I've downloaded and installed 20.11 on my `x86_64` workstation, but haven't configured the Lmod modules yet that I wanna use with it -- things like CUDA 11.1, UCX, OpenMPI 4, and OpenBLAS 0.3.12 or newer. Unless 20.11 actually already bundles OpenMPI 4 in addition to the default OpenMPI 3 @cparrott73 ?

Edit 2: Some details: with both 20.7 and 20.9, it complains about `-march=native`:
```
$ make clean
nvc-Error-Unknown switch: -march=native
make: *** [Makefile.prebuild:66: getarch] Error 1
Makefile.system:1279: Makefile.: No such file or directory
make: *** No rule to make target 'Makefile.'.  Stop.
```
```
$ nvc --version

nvc 20.7-0 LLVM 64-bit target on x86-64 Linux -tp zen 
NVIDIA Compilers and Tools
Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
```
$ nvc --version

nvc 20.9-0 LLVM 64-bit target on x86-64 Linux -tp zen 
NVIDIA Compilers and Tools
Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```

> "cc clobber ignored"

@martin-frbg  Are you referring to [this link](https://stackoverflow.com/questions/64808014/avoid-warning-cc-clobber-ignored-with-pgi-on-running-on-valgrind)? Yeah, I got that warning too with 20.11. For me, with 0.3.13 (after patching with @cparrott73 's patch above in the OP) and NVIDIA HPC SDK 20.11, the build failed at creating the `.a` file using `ranlib` (but that might be from my imperfect self-maintained Lmod module files)
```
ar: `u' modifier ignored since `D' is the default (see `U')
/opt/nvidia/hpc_sdk/Linux_x86_64/20.11/compilers/bin/ranlib ../../libopenblas_zenp-r0.3.13.a
make[2]: /opt/nvidia/hpc_sdk/Linux_x86_64/20.11/compilers/bin/ranlib: Command not found
make[2]: *** [Makefile:601: ../../libopenblas_zenp-r0.3.13.a] Error 127
make[2]: Leaving directory '/home/wyp/compile/OpenBLAS-0.3.13/lapack-netlib/SRC'
make[1]: *** [Makefile:27: lapacklib] Error 2
make[1]: Leaving directory '/home/wyp/compile/OpenBLAS-0.3.13/lapack-netlib'
make: *** [Makefile:244: netlib] Error 2
```
Here's the full build log, in case you wanna take a closer look.
[build-nv20.11.log.gz](https://github.com/xianyi/OpenBLAS/files/5698366/build-nv20.11.log.gz)

I wonder if this `ranlib` issue is related to issue #3037?

Also, I just finished building Reference LAPACK 3.9.0 from the tarball on Netlib on this `x86_64` machine (henceforth I will refer to as BaseCamp). The `-Kieee` flag is required to avoid hundreds of numerical errors, even for reference LAPACK. Full details here: https://github.com/Reference-LAPACK/lapack/issues/430

> Does the same patch apply to 20.7 or 20.9?
> 
> Edit: I've downloaded and installed 20.11 on my `x86_64` workstation, but haven't configured the Lmod modules yet that I wanna use with it -- things like CUDA 11.1, UCX, OpenMPI 4, and OpenBLAS 0.3.12 or newer. Unless 20.11 actually already bundles OpenMPI 4 in addition to the default OpenMPI 3 @cparrott73 ?
> 
> Edit 2: Some details: with both 20.7 and 20.9, it complains about `-march=native`:
> 
> ```
> $ make clean
> nvc-Error-Unknown switch: -march=native
> make: *** [Makefile.prebuild:66: getarch] Error 1
> Makefile.system:1279: Makefile.: No such file or directory
> make: *** No rule to make target 'Makefile.'.  Stop.
> ```
> 
> ```
> $ nvc --version
> 
> nvc 20.7-0 LLVM 64-bit target on x86-64 Linux -tp zen 
> NVIDIA Compilers and Tools
> Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
> ```
> 
> ```
> $ nvc --version
> 
> nvc 20.9-0 LLVM 64-bit target on x86-64 Linux -tp zen 
> NVIDIA Compilers and Tools
> Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
> ```

Unfortunately, I don't think this patch will work for 20.7 or 20.9.  Several bugs that were causing some compilation errors with some of the kernels that use inline assembly got fixed for 20.11.  You can certainly try, but I would highly recommend updating to 20.11 if possible.

> Does the same patch apply to 20.7 or 20.9?
> 
> Edit: I've downloaded and installed 20.11 on my `x86_64` workstation, but haven't configured the Lmod modules yet that I wanna use with it -- things like CUDA 11.1, UCX, OpenMPI 4, and OpenBLAS 0.3.12 or newer. Unless 20.11 actually already bundles OpenMPI 4 in addition to the default OpenMPI 3 @cparrott73 ?
> 
> Edit 2: Some details: with both 20.7 and 20.9, it complains about `-march=native`:
> 
> ```
> $ make clean
> nvc-Error-Unknown switch: -march=native
> make: *** [Makefile.prebuild:66: getarch] Error 1
> Makefile.system:1279: Makefile.: No such file or directory
> make: *** No rule to make target 'Makefile.'.  Stop.
> ```
> 
> ```
> $ nvc --version
> 
> nvc 20.7-0 LLVM 64-bit target on x86-64 Linux -tp zen 
> NVIDIA Compilers and Tools
> Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
> ```
> 
> ```
> $ nvc --version
> 
> nvc 20.9-0 LLVM 64-bit target on x86-64 Linux -tp zen 
> NVIDIA Compilers and Tools
> Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
> ```

Missed a comment above...we actually do bundle Open MPI 4 with 20.11, but it is not default.  It is considered a "beta" right now, built against UCX 1.9.0.  You are certainly welcome to give it a go and see how it fares, although it is entirely possible that there could be issues.  It seems to work well in my testing so far, but we are still putting it through its paces.

Using `-tp skylake` both crashes the build and testing phase with `Illegal instruction`, which is kinda counterintuitive? Unless not all Skylake processors are created equal... It's just a laptop Core i5 processor, after all, not a Xeon processor.
```
OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 ./sblat2 < ./sblat2.dat
Illegal instruction (core dumped)
make[1]: *** [Makefile:28: level1] Error 132
make[1]: *** Waiting for unfinished jobs....
Illegal instruction (core dumped)
make[1]: *** [Makefile:88: level2] Error 132
```

Full build and testing logs:
[build-nv20.11-skylake.log.gz](https://github.com/xianyi/OpenBLAS/files/5704289/build-nv20.11-skylake.log.gz)
[test-nv20.11-skylake.log.gz](https://github.com/xianyi/OpenBLAS/files/5704296/test-nv20.11-skylake.log.gz)

> Using `-tp skylake` both crashes the build and testing phase with `Illegal instruction`, which is kinda counterintuitive? Unless not all Skylake processors are created equal... It's just a laptop Core i5 processor, after all, not a Xeon processor.
> 
> ```
> OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 ./sblat2 < ./sblat2.dat
> Illegal instruction (core dumped)
> make[1]: *** [Makefile:28: level1] Error 132
> make[1]: *** Waiting for unfinished jobs....
> Illegal instruction (core dumped)
> make[1]: *** [Makefile:88: level2] Error 132
> ```
> 
> Full build and testing logs:
> [build-nv20.11-skylake.log.gz](https://github.com/xianyi/OpenBLAS/files/5704289/build-nv20.11-skylake.log.gz)
> [test-nv20.11-skylake.log.gz](https://github.com/xianyi/OpenBLAS/files/5704296/test-nv20.11-skylake.log.gz)

@wyphan - our compilers will report what CPU they detect on your system if you pass the `-V` option, e.g.

```
$ nvc -V

nvc 20.11-0 LLVM 64-bit target on x86-64 Linux -tp skylake 
NVIDIA Compilers and Tools
Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```

Note the `-tp skylake` here.

It sounds like you built OpenBLAS for a CPU architecture that your i5 doesn't support.  Worth noting that not all "Skylakes" are created equal.  `-tp skylake` for our compilers means a Xeon-class CPU that supports AVX-512.  A lot of consumer-grade Intel CPUs that are called "Skylake" or later do not actually support AVX-512, and so our compilers will actually report those as `-tp haswell` instead.  I would suggest changing `-tp skylake` to `-tp haswell` in your makefile, and rebuilding.  That will probably work for you.

With `-tp p7-64` and default optimization level, the build crashes here for me:
```
/opt/nvidia/hpc_sdk/Linux_x86_64/20.11/compilers/bin/nvc -c -O2 -DMAX_STACK_ALLOC=2048 -mp -tp p7-64 -DF_INTERFACE_PGI -fPIC -DSMP_SERVER -DUSE_OPENMP -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.12\" -msse3 -mssse3 -msse4.1 -DASMNAME=sdot_k -DASMFNAME=sdot_k_ -DNAME=sdot_k_ -DCNAME=sdot_k -DCHAR_NAME=\"sdot_k_\" -DCHAR_CNAME=\"sdot_k\" -DNO_AFFINITY -I.. -UDOUBLE  -UCOMPLEX -UCOMPLEX -UDOUBLE ../kernel/x86_64/sdot.c -o sdot_k.o
"../common_x86_64.h", line 163: warning: "cc" clobber ignored
      ("cpuid": "=a" (*eax), "=b" (*ebx), "=c" (*ecx), "=d" (*edx) : "0" (op), "2" (count) : "cc");
                                                                                             ^

"../common_x86_64.h", line 163: warning: "cc" clobber ignored
      ("cpuid": "=a" (*eax), "=b" (*ebx), "=c" (*ecx), "=d" (*edx) : "0" (op), "2" (count) : "cc");
                                                                                             ^

"../kernel/x86_64/saxpy_microk_haswell-2.c", line 69: warning: "cc" clobber
          ignored
  	: "cc", 
  	  ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 6: error: identifier
          "__m256" is undefined
  typedef __m256  v_f32;
          ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 7: error: identifier
          "__m256d" is undefined
  typedef __m256d  v_f64;
          ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 31: error: identifier
          "__m256" is undefined
  BLAS_FINLINE float v_sum_f32(__m256 a)
                               ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 33: error: identifier
          "__m256" is undefined
      __m256 sum_halves = _mm256_hadd_ps(a, a);
      ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 33: warning: function
          "_mm256_hadd_ps" declared implicitly
      __m256 sum_halves = _mm256_hadd_ps(a, a);
                          ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 35: warning: function
          "_mm256_castps256_ps128" declared implicitly
      __m128 lo = _mm256_castps256_ps128(sum_halves);
                  ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 35: error: a value of type
          "int" cannot be used to initialize an entity of type "__m128"
      __m128 lo = _mm256_castps256_ps128(sum_halves);
                  ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 36: warning: function
          "_mm256_extractf128_ps" declared implicitly
      __m128 hi = _mm256_extractf128_ps(sum_halves, 1);
                  ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 36: error: a value of type
          "int" cannot be used to initialize an entity of type "__m128"
      __m128 hi = _mm256_extractf128_ps(sum_halves, 1);
                  ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 41: error: identifier
          "__m256d" is undefined
  BLAS_FINLINE double v_sum_f64(__m256d a)
                                ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 43: error: identifier
          "__m256d" is undefined
      __m256d sum_halves = _mm256_hadd_pd(a, a);
      ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 43: warning: function
          "_mm256_hadd_pd" declared implicitly
      __m256d sum_halves = _mm256_hadd_pd(a, a);
                           ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 44: warning: function
          "_mm256_castpd256_pd128" declared implicitly
      __m128d lo = _mm256_castpd256_pd128(sum_halves);
                   ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 44: error: a value of type
          "int" cannot be used to initialize an entity of type "__m128d"
      __m128d lo = _mm256_castpd256_pd128(sum_halves);
                   ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 45: warning: function
          "_mm256_extractf128_pd" declared implicitly
      __m128d hi = _mm256_extractf128_pd(sum_halves, 1);
                   ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 45: error: a value of type
          "int" cannot be used to initialize an entity of type "__m128d"
      __m128d hi = _mm256_extractf128_pd(sum_halves, 1);
                   ^

"../kernel/x86_64/../arm/../simd/intrin_avx.h", line 60: warning: last line of
          file ends without a newline
  #define v_zero_f64 _mm256_setzero_pd
                                      ^

"../kernel/x86_64/../arm/sum.c", line 69: warning: function "_mm256_setzero_ps"
          declared implicitly
  		v_f32 vsum0 = v_zero_f32();
  		              ^

"../kernel/x86_64/../arm/sum.c", line 75: warning: function "_mm256_add_ps"
          declared implicitly
  			vsum0 = v_add_f32(vsum0, v_loadu_f32(x));
  			        ^

"../kernel/x86_64/../arm/sum.c", line 75: warning: function "_mm256_loadu_ps"
          declared implicitly
  			vsum0 = v_add_f32(vsum0, v_loadu_f32(x));
  			                         ^

"../kernel/x86_64/../arm/sum.c", line 81: warning: function "_mm256_add_ps"
          declared implicitly
  		vsum0 = v_add_f32(
  		        ^

"../kernel/x86_64/../arm/sum.c", line 85: warning: function "_mm256_loadu_ps"
          declared implicitly
  			vsum0 = v_add_f32(vsum0, v_loadu_f32(x + i));
  			                         ^

10 errors detected in the compilation of "../kernel/x86_64/../arm/sum.c".
make[1]: *** [Makefile.L1:691: ssum_k.o] Error 2
```
Full build log:
[build-nv20.11-p7-64.log.gz](https://github.com/xianyi/OpenBLAS/files/5705907/build-nv20.11-p7-64.log.gz)

--- a/kernel/power/KERNEL.POWER10
+++ b/kernel/power/KERNEL.POWER10
@@ -169,8 +169,13 @@ ZROTKERNEL   = zrot.c
 #
 SSCALKERNEL  = sscal.c
 DSCALKERNEL  = dscal.c
+ifeq ($(C_COMPILER), PGI)
+CSCALKERNEL  = ../arm/zscal.c
+ZSCALKERNEL  = ../arm/zscal.c
+else
 CSCALKERNEL  = zscal.c
 ZSCALKERNEL  = zscal.c
+endif
 #
 SSWAPKERNEL  = sswap.c
 DSWAPKERNEL  = dswap.c

--- a/kernel/power/KERNEL.POWER8
+++ b/kernel/power/KERNEL.POWER8
@@ -242,8 +242,13 @@ ZROTKERNEL   = zrot.c
 #
 SSCALKERNEL  = sscal.c
 DSCALKERNEL  = dscal.c
+ifeq ($(C_COMPILER), PGI)
+CSCALKERNEL  = ../arm/zscal.c
+ZSCALKERNEL  = ../arm/zscal.c
+else
 CSCALKERNEL  = zscal.c
 ZSCALKERNEL  = zscal.c
+endif
 #
 SSWAPKERNEL  = sswap.c
 DSWAPKERNEL  = dswap.c

--- a/kernel/power/KERNEL.POWER9
+++ b/kernel/power/KERNEL.POWER9
@@ -166,8 +166,13 @@ ZROTKERNEL   = zrot.c
 #
 SSCALKERNEL  = sscal.c
 DSCALKERNEL  = dscal.c
+ifeq ($(C_COMPILER), PGI)
+CSCALKERNEL  = ../arm/zscal.c
+ZSCALKERNEL  = ../arm/zscal.c
+else
 CSCALKERNEL  = zscal.c
 ZSCALKERNEL  = zscal.c
+endif
 #
 SSWAPKERNEL  = sswap.c
 DSWAPKERNEL  = dswap.c

